cmd/kubeapps-apis/plugins/kapp_controller/packages/v1alpha1/server_ctrl_packages.go

```diff
@@ -285,91 +284,155 @@ func (s *Server) GetInstalledPackageSummaries(ctx context.Context, request *core
-
-	// TODO(agamez): DRY up this logic (cf GetAvailablePackageSummaries)
 	if len(pkgInstalls) > 0 {
 		startAt := -1
 		if pageSize > 0 {
 			startAt = int(pageSize) * pageOffset
+			if startAt > len(pkgInstalls) {
+				return nil, status.Errorf(codes.InvalidArgument, "invalid pagination arguments %v", request.GetPaginationOptions())
+			}
+			pkgInstalls = pkgInstalls[startAt:]
+			if len(pkgInstalls) > int(pageSize) {
+				pkgInstalls = pkgInstalls[:pageSize]
+			}
 		}
-		for i, pkgInstall := range pkgInstalls {
-			wg.Add(1)
-			if startAt <= i {
-				go func(i int, pkgInstall *packagingv1alpha1.PackageInstall) error {
-					defer wg.Done()
-					// fetch additional information from the package metadata
-					// TODO(agamez): if the repository where the package belongs to is not installed, it will throw an error;
-					// decide whether we should ignore it or it is ok with returning an error
-					pkgMetadata, err := s.getPkgMetadata(ctx, cluster, pkgInstall.ObjectMeta.Namespace, pkgInstall.Spec.PackageRef.RefName)
-					if err != nil {
-						return statuserror.FromK8sError("get", "PackageMetadata", pkgInstall.Spec.PackageRef.RefName, err)
-					}
 
-					// Use the field selector to return only Package CRs that match on the spec.refName.
-					fieldSelector := fmt.Sprintf("spec.refName=%s", pkgInstall.Spec.PackageRef.RefName)
-					pkgs, err := s.getPkgsWithFieldSelector(ctx, cluster, pkgInstall.ObjectMeta.Namespace, fieldSelector)
-					if err != nil {
-						return statuserror.FromK8sError("get", "Package", "", err)
-					}
-					pkgVersionsMap, err := getPkgVersionsMap(pkgs)
-					if err != nil {
-						return err
-					}
-
-					// generate the installedPackageSummary from the fetched information
-					installedPackageSummary, err := s.buildInstalledPackageSummary(pkgInstall, pkgMetadata, pkgVersionsMap, cluster)
-					if err != nil {
-						return status.Errorf(codes.Internal, fmt.Sprintf("unable to create the InstalledPackageSummary: %v", err))
+		// Collect a set of all package names with their corresponding package
+		// metadata and package version data.
+		// Complication 1: PackageMetadata and Package resources are not CRs,
+		// but rather aggregated API resources, which means a PackageInstall in
+		// namespace "default" may refer to a Package (and metadata) in the same
+		// "default" namespace, or from the global namespace.
+		// Complication 2: PackageMetadata and Package resources for can be
+		// present in different namespaces with the same name and different
+		// versions etc.
+		// With this in mind, we create a data structure that records the
+		// metadata and versions per package refName per namespace.
+		// TODO(minelson) move this struct initialization out.
+		type pkgMetaAndVersionsData struct {
+			meta     *datapackagingv1alpha1.PackageMetadata
+			versions map[string]*datapackagingv1alpha1.Package
+		}
+		pkgDatas := make(map[string]map[string]*pkgMetaAndVersionsData)
+		for _, pi := range pkgInstalls {
```

Comment:
Yeah, I do too... I think I initially was having naming issues and thought this would just be used over 2 lines, but it's not. I'll update before landing.